src/System.Management.Automation/engine/hostifaces/History.cs

```diff
@@ -1041,114 +1019,137 @@ public string Id
+                // Get the history entry to invoke
+                HistoryInfo entry = GetHistoryEntryToInvoke(history);
+                string commandToInvoke = entry.CommandLine;
 
-                EventHandler<DataAddedEventArgs> debugAdded = delegate (object sender, DataAddedEventArgs e) { DebugRecord record = (DebugRecord)((PSDataCollection<DebugRecord>)sender)[e.Index]; WriteDebug(record.Message); };
-                EventHandler<DataAddedEventArgs> errorAdded = delegate (object sender, DataAddedEventArgs e) { ErrorRecord record = (ErrorRecord)((PSDataCollection<ErrorRecord>)sender)[e.Index]; WriteError(record); };
-                EventHandler<DataAddedEventArgs> informationAdded = delegate (object sender, DataAddedEventArgs e) { InformationRecord record = (InformationRecord)((PSDataCollection<InformationRecord>)sender)[e.Index]; WriteInformation(record); };
-                EventHandler<DataAddedEventArgs> progressAdded = delegate (object sender, DataAddedEventArgs e) { ProgressRecord record = (ProgressRecord)((PSDataCollection<ProgressRecord>)sender)[e.Index]; WriteProgress(record); };
-                EventHandler<DataAddedEventArgs> verboseAdded = delegate (object sender, DataAddedEventArgs e) { VerboseRecord record = (VerboseRecord)((PSDataCollection<VerboseRecord>)sender)[e.Index]; WriteVerbose(record.Message); };
-                EventHandler<DataAddedEventArgs> warningAdded = delegate (object sender, DataAddedEventArgs e) { WarningRecord record = (WarningRecord)((PSDataCollection<WarningRecord>)sender)[e.Index]; WriteWarning(record.Message); };
+                if (!ShouldProcess(commandToInvoke))
+                {
+                    return;
+                }
 
-                ps.Streams.Debug.DataAdded += debugAdded;
-                ps.Streams.Error.DataAdded += errorAdded;
-                ps.Streams.Information.DataAdded += informationAdded;
-                ps.Streams.Progress.DataAdded += progressAdded;
-                ps.Streams.Verbose.DataAdded += verboseAdded;
-                ps.Streams.Warning.DataAdded += warningAdded;
+                // Check if there is a loop in invoke-history
+                if (history.PresentInInvokeHistoryEntrySet(entry))
+                {
+                    ThrowTerminatingError(
+                        new ErrorRecord(
+                            new InvalidOperationException(HistoryStrings.InvokeHistoryLoopDetected),
+                            "InvokeHistoryLoopDetected",
+                            ErrorCategory.InvalidOperation,
+                            targetObject: null));
+                }
+                else
+                {
+                    history.AddToInvokeHistoryEntrySet(entry);
+                }
 
-                LocalRunspace localRunspace = ps.Runspace as LocalRunspace;
+                // Replace Invoke-History with string which is getting invoked
+                ReplaceHistoryString(entry, ctxRunspace);
```

Comment:
Today `Invoke-Command` will replace the current history with the one it's about invoke. That behavior makes sense when it only supports the invocation of one history entry.

After the change, `Invoke-Command` can invoke multiple command entries, and thus the current history will be replaced with the last command entry. The behavior doesn't make sense IMO when invoke multiple histories. Maybe in case of invoking multiple history entries, we shouldn't change the history string.

Reply:
Yeah, I reckon it should just leave the Invoke-History line itself in the history, there isn't really a better way to capture it (unless you wanted to concatenate _all_ the invoked history entries and combine them into one history entry, but I don't think that's really necessary)